src/apps/MASS3DPA-StdPar.cpp

```diff
@@ -0,0 +1,230 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-21, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "MASS3DPA.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#include <ranges>
+#include <algorithm>
+#include <execution>
+
+#include <iostream>
+
+namespace rajaperf {
+namespace apps {
+
+//#define USE_RAJA_UNROLL
+#define RAJA_DIRECT_PRAGMA(X) _Pragma(#X)
+#if defined(USE_RAJA_UNROLL)
+#define RAJA_UNROLL(N) RAJA_DIRECT_PRAGMA(unroll(N))
+#else
+#define RAJA_UNROLL(N)
+#endif
+#define FOREACH_THREAD(i, k, N) for (int i = 0; i < N; i++)
+
+void MASS3DPA::runStdParVariant(VariantID vid) {
+#if defined(RUN_STDPAR)
+
+  const Index_type run_reps = getRunReps();
+
+  MASS3DPA_DATA_SETUP;
+
+  switch (vid) {
+
+  case Base_StdPar: {
+
+    auto range = std::views::iota(0,(int)NE);
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      std::for_each( std::execution::par_unseq,
+                      std::begin(range), std::end(range),
+                      [=](int e) {
+
+        MASS3DPA_0_CPU
+
+        FOREACH_THREAD(dy, y, D1D) {
+          FOREACH_THREAD(dx, x, D1D){
+            MASS3DPA_1
+          }
+          FOREACH_THREAD(dx, x, Q1D) {
+            MASS3DPA_2
+          }
+        }
+
+        FOREACH_THREAD(dy, y, D1D) {
+          FOREACH_THREAD(qx, x, Q1D) {
+            MASS3DPA_3
+          }
+        }
+
+        FOREACH_THREAD(qy, y, Q1D) {
+          FOREACH_THREAD(qx, x, Q1D) {
+            MASS3DPA_4
+          }
+        }
+
+        FOREACH_THREAD(qy, y, Q1D) {
+          FOREACH_THREAD(qx, x, Q1D) {
+            MASS3DPA_5
+          }
+        }
+
+        FOREACH_THREAD(d, y, D1D) {
+          FOREACH_THREAD(q, x, Q1D) {
+            MASS3DPA_6
+          }
+        }
+
+        FOREACH_THREAD(qy, y, Q1D) {
+          FOREACH_THREAD(dx, x, D1D) {
+            MASS3DPA_7
+          }
+        }
+
+        FOREACH_THREAD(dy, y, D1D) {
+          FOREACH_THREAD(dx, x, D1D) {
+            MASS3DPA_8
+          }
+        }
+
+        FOREACH_THREAD(dy, y, D1D) {
+          FOREACH_THREAD(dx, x, D1D) {
+            MASS3DPA_9
+          }
+        }
+
+      }); // element loop
+    }
+    stopTimer();
+
+    break;
+  }
+
+#if defined(RUN_RAJA_STDPAR)
+  case RAJA_StdPar: {
+
+    //Currently Teams requires two policies if compiled with a device
+    using launch_policy = RAJA::expt::LaunchPolicy<RAJA::expt::seq_launch_t
+#if defined(RAJA_DEVICE_ACTIVE)
+                                                   ,m3d_device_launch
+#endif
+                                                   >;
+
+    using outer_x = RAJA::expt::LoopPolicy<RAJA::loop_exec
```

Comment:
@jeffhammond will there be companion RAJA backed for this work? Otherwise comparing with the RAJA versions (which look to run serial) won't be quite fair.

Reply:
initially, i was replacing the names as if there was going to be a RAJA stdpar back-end, but then i got lazy.  in any case, all of this code is guarded by macros so it won't be touched.

i am not committing to doing a RAJA back-end for stdpar but i suppose i'll think about it.  my hope was that you all would see how much better C++20 is than OpenMP and implement it yourselves 😄

Reply:
Thanks @jeffhammond. We'll discuss how we want to move forward with this. We're in the middle of a very hectic period for the next month or so, so it will probably be a few weeks before we can get back to you with anything definitive. We're moving to requiring C++14 for RAJA with the next release (November). It will be a while before we can expect user apps to rely on C++20. Production codes are like oil tankers and take some time and space to change course 😄 . Nevertheless, I think this contribution will be useful because it will allow us to assess whether compilers are seeking out certain patterns and constructs and giving less support for other language features.

Reply:
C++20 is only required when you use it for parallelism so this isn't disruptive to the project. It's not unlike how CUDA or OpenMP compiler support is required when you use it, but doesn't break things that don't.

In any case, I am in no hurry here and look forward to further discussion when you all have the time.